File: src/cpp/src/operations/data_average.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// Point <-> cell data averaging. The point->cell direction is a pure gather
// (one output row per cell, reading only that cell's nodes) and is therefore
// parallel over cells. The cell->point direction is a scatter, so its
// accumulation pass is deliberately SERIAL: floating-point addition is not
// associative, and a parallel scatter would make the result depend on the
// thread count and on the scheduling, which would flake the cross-backend
// tests. Only the final divide is parallelised. This is the same phase-split
// philosophy as operations/surface.cpp.
//
// Geometry is never modified. See operations/data_average.hpp for the contract.

// System includes
#include <algorithm>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <memory>
#include <stdexcept>
#include <string>
#include <unordered_set>
#include <utility>
#include <vector>

// Project includes
#include "meshioplusplus/operations/data_average.hpp"
#include "meshioplusplus/detail/data_ops.hpp"
#include "meshioplusplus/detail/value_io.hpp"
#include "meshioplusplus/log.hpp"
#include "meshioplusplus/ndarray.hpp"
#include "meshioplusplus/operations/data_common.hpp"
#include "meshioplusplus/parallel.hpp"

// Project includes (private, not installed)
#include "../detail/slot_runs.hpp"
#include "../detail/typed_view.hpp"

namespace meshioplusplus {

namespace {

/// Output shape for `rows` entries of `ncomp` components: 1-D when scalar, so
/// scalar data stays scalar (the same rule merge/clean follow).
std::vector<std::size_t> davg_shape(std::size_t rows, std::size_t ncomp) {
    if (ncomp <= 1)
        return {rows};
    return {rows, ncomp};
}

/// Applies the NaN policy to one produced value.
double davg_apply_nan_policy(double v, const DataAverageOptions& rOpts, const char* pWhat,
                             const std::string& rName, std::size_t index) {
    if (std::isfinite(v))
        return v;
    switch (rOpts.nan_policy) {
        case NanPolicy::Ignore:
            return v;
        case NanPolicy::Replace:
            return rOpts.nan_replacement;
        case NanPolicy::Fail:
            throw std::invalid_argument(
                std::string("meshio++: data_average: non-finite value for ") + pWhat + " " +
                std::to_string(index) + " of '" + rName + "'");
    }
    return v;
}

/// The target name for one input array, per prefix/suffix.
std::string davg_target_name(const std::string& rName, const DataAverageOptions& rOpts) {
    return rOpts.prefix + rName + rOpts.suffix;
}

/// Resolves the requested names, validating each against the source location.
std::vector<std::string> davg_resolve_names(const Mesh& rMesh, DataLocation source,
                                            const DataAverageOptions& rOpts) {
    if (rOpts.names.empty())
        return data_names(rMesh, source);
    for (const std::string& n : rOpts.names)
        if (!data_has(rMesh, source, n))
            throw std::invalid_argument(data_unknown_key_message(rMesh, source, n));
    return rOpts.names;
}

/// Rejects an output name that already exists when overwriting is disabled.
void davg_check_target(const Mesh& rMesh, DataLocation target, const std::string& rTarget,
                       const DataAverageOptions& rOpts) {
    if (!rOpts.overwrite && data_has(rMesh, target, rTarget))
        throw std::invalid_argument(std::string("meshio++: data_average: ") +
                                    data_location_name(target) + " '" + rTarget +
                                    "' already exists (pass overwrite=true to replace it)");
}

/// The node ids of one cell, covering rectangular, polygon and polyhedron
/// blocks. For a polyhedron the *distinct* nodes across all faces are used.
void davg_cell_nodes(const Mesh::CellView& rCell, std::size_t index,
                     std::vector<std::int64_t>& rNodes) {
    rNodes.clear();
    if (rCell.IsPolyhedron()) {
        std::unordered_set<std::int64_t> seen;
        for (std::size_t f = 0; f < rCell.NumFaces(index); ++f) {
            const auto face = rCell.Face(index, f);
            for (std::size_t i = 0; i < face.second; ++i)
                if (seen.insert(face.first[i]).second)
                    rNodes.push_back(face.first[i]);
        }
        return;
    }
    if (rCell.IsRagged()) {
        const std::int64_t* row = rCell.Row(index);
        rNodes.assign(row, row + rCell.RowSize(index));
        return;
    }
    // Connectivity dtype is not guaranteed to be Int64 (a MESHIO-backed mesh
    // often carries Int32 straight from numpy), so read it dtype-agnostically.
    const std::size_t npc = rCell.NodesPerCell();
    const NDArray& conn = rCell.Conn();
    rNodes.reserve(npc);
    for (std::size_t k = 0; k < npc; ++k)
        rNodes.push_back(detail::read_int(conn, index * npc + k));
}

/// Every point's incident cells (global, block-major ids), one entry per
/// occurrence of the point in `davg_cell_nodes`, in (block, cell, node)
/// order: the order in which the serial scatter added their contributions.
/// Entries naming no point of the mesh are dropped, as the scatter skipped
/// them.
struct DavgIncidence {
    std::vector<std::uint64_t> mStart;   ///< `npoints + 1` offsets into `mCellOf`
    std::vector<std::uint64_t> mCellOf;  ///< per entry, its global cell
};

DavgIncidence davg_incidence(const Mesh& rMesh, const std::vector<std::size_t>& rBlockBase,
                             std::size_t NumPoints) {
    const std::size_t nblocks = rMesh.NumCellBlocks();
    const std::size_t ncells = rBlockBase[nblocks];
    std::vector<std::uint64_t> count(ncells);
    for (std::size_t b = 0; b < nblocks; ++b) {
        const auto cb = rMesh.Cells(b);
        parallel_for(cb.NumCells(), [&](std::size_t c) {
            thread_local std::vector<std::int64_t> nodes;
            davg_cell_nodes(cb, c, nodes);
            count[rBlockBase[b] + c] = nodes.size();
        });
    }
    std::vector<std::uint64_t> offset(ncells);
    const std::uint64_t nentries =
        parallel_exclusive_scan(count.data(), ncells, offset.data(), std::uint64_t{0});
    std::vector<std::uint64_t> point_of(nentries);
    std::vector<std::uint64_t> cell_of(nentries);
    for (std::size_t b = 0; b < nblocks; ++b) {
        const auto cb = rMesh.Cells(b);
        parallel_for(cb.NumCells(), [&](std::size_t c) {
            thread_local std::vector<std::int64_t> nodes;
            davg_cell_nodes(cb, c, nodes);
            const std::size_t gc = rBlockBase[b] + c;
            std::uint64_t e = offset[gc];
            for (std::int64_t node : nodes) {
                point_of[e] = node < 0 || static_cast<std::size_t>(node) >= NumPoints
                                  ? NumPoints
                                  : static_cast<std::uint64_t>(node);
                cell_of[e] = gc;
                ++e;
            }
        });
    }
    // A stable counting sort by point keeps each point's entries in entry
    // order; bucket `NumPoints` collects the dropped ones.
    std::vector<std::uint64_t> order;
    DavgIncidence out;
    detail::slot_runs_impl::counting_sort(
        nentries, NumPoints + 1, [&](std::uint64_t e) { return point_of[e]; }, order, out.mStart);
    out.mStart.pop_back();  // drop the invalid bucket's end
    const std::uint64_t kept = out.mStart[NumPoints];
    out.mCellOf.resize(kept);
    parallel_for_bw(kept, [&](std::size_t j) { out.mCellOf[j] = cell_of[order[j]]; });
    return out;
}

}  // namespace

CellPointWeight cell_point_weight_from_name(const std::string& rName) {
    if (rName == "uniform" || rName == "simple")
        return CellPointWeight::Uniform;
    if (rName == "measure" || rName == "area" || rName == "volume")
        return CellPointWeight::Measure;
    throw std::invalid_argument("meshio++: unknown averaging weight '" + rName +
                                "' (expected 'uniform' or 'measure')");
}

Mesh point_data_to_cell_data(const Mesh& rMesh, const DataAverageOptions& rOpts) {
    const std::vector<std::string> names = davg_resolve_names(rMesh, DataLocation::Point, rOpts);
    for (const std::string& n : names)
        davg_check_target(rMesh, DataLocation::Cell, davg_target_name(n, rOpts), rOpts);

    Mesh out = detail::clone_mesh(rMesh);
    const std::size_t nblocks = rMesh.NumCellBlocks();
    if (nblocks == 0)
        return out;

    for (const std::string& name : names) {
        const NDArray& src = rMesh.PointData(name);
        const std::size_t ncomp = data_num_components(src);
        const std::size_t npoints = rMesh.NumPoints();

        // One output array per cell block — the uniform API requires exactly
        // NumCellBlocks() of them.
        std::vector<NDArray> blocks;
        blocks.reserve(nblocks);
        for (std::size_t b = 0; b < nblocks; ++b) {
            const auto cb = rMesh.Cells(b);
            const std::size_t nc = cb.NumCells();
            NDArray dst(DType::Float64, davg_shape(nc, ncomp));
            double* pdst = dst.As<double>();

            // Pure gather: each cell reads only its own nodes, so this is safe
            // to run in parallel and is genuine per-element compute.
            parallel_for(nc, [&](std::size_t c) {
                std::vector<std::int64_t> nodes;
                davg_cell_nodes(cb, c, nodes);
                for (std::size_t k = 0; k < ncomp; ++k) {
                    double sum = 0.0;
                    std::int64_t count = 0;
                    for (std::int64_t node : nodes) {
                        if (node < 0 || static_cast<std::size_t>(node) >= npoints)
                            continue;
                        const double v =
                            detail::read_double(src, static_cast<std::size_t>(node) * ncomp + k);
                        if (std::isfinite(v)) {
                            sum += v;
                            ++count;
                        }
                    }
                    pdst[c * ncomp + k] =
                        count > 0 ? sum / static_cast<double>(count) : std::nan("");
                }
            });

            // The NaN policy is applied serially afterwards so that a `Fail`
            // throw is deterministic rather than racing between threads.
            for (std::size_t i = 0; i < nc * ncomp; ++i)
                pdst[i] =
                    davg_apply_nan_policy(pdst[i], rOpts, "cell", name, i / (ncomp ? ncomp : 1));
            blocks.push_back(std::move(dst));
        }
        out.AddCellData(davg_target_name(name, rOpts), std::move(blocks));
    }
    return out;
}

Mesh cell_data_to_point_data(const Mesh& rMesh, const DataAverageOptions& rOpts) {
    const std::vector<std::string> names = davg_resolve_names(rMesh, DataLocation::Cell, rOpts);
    for (const std::string& n : names)
        davg_check_target(rMesh, DataLocation::Point, davg_target_name(n, rOpts), rOpts);

    Mesh out = detail::clone_mesh(rMesh);
    const std::size_t npoints = rMesh.NumPoints();
    const std::size_t nblocks = rMesh.NumCellBlocks();
    if (npoints == 0 || nblocks == 0)
        return out;

    // Per-cell weights, computed once and reused by every array.
    std::vector<std::vector<double>> weights(nblocks);
    bool warned_measure = false;
    for (std::size_t b = 0; b < nblocks; ++b) {
        const auto cb = rMesh.Cells(b);
        weights[b].assign(cb.NumCells(), 1.0);
        if (rOpts.weight != CellPointWeight::Measure)
            continue;
        const NDArray& points = rMesh.Points();
        const std::size_t pdim = rMesh.PointDim();
        std::vector<double>& w = weights[b];
        parallel_for(cb.NumCells(),
                     [&](std::size_t c) { w[c] = detail::cell_measure(points, pdim, cb, c); });
        for (double& v : w) {
            if (!std::isfinite(v) || v <= 0.0) {
                v = 1.0;
                if (!warned_measure) {
                    warned_measure = true;
                    log::warn(
                        "data_average: cell measure unavailable for some cells (ragged or "
                        "unsupported type); falling back to a unit weight there");
                }
            }
        }
    }

    // Built on first use and shared by every array.
    std::unique_ptr<DavgIncidence> incidence;
    std::vector<double> cell_weight;
    std::vector<std::size_t> cell_block_base;
    for (const std::string& name : names) {
        if (rMesh.CellDataNumBlocks(name) != nblocks)
            throw std::invalid_argument("meshio++: data_average: cell_data '" + name + "' has " +
                                        std::to_string(rMesh.CellDataNumBlocks(name)) +
                                        " block(s) but the mesh has " + std::to_string(nblocks) +
                                        " cell block(s)");

        const std::size_t ncomp = data_num_components(rMesh.CellData(name, 0));
        for (std::size_t b = 0; b < nblocks; ++b)
            if (data_num_components(rMesh.CellData(name, b)) != ncomp)
                throw std::invalid_argument("meshio++: data_average: cell_data '" + name +
                                            "' has inconsistent component counts across blocks");
        NDArray dst(DType::Float64, davg_shape(npoints, ncomp));
        double* pdst = dst.As<double>();
        if (detail::slot_runs_impl::workers() <= 1) {
            // One worker (a SEQ build, or one thread): the plain scatter, which
            // adds the same terms in the same order as the gather below and is
            // cheaper than building the incidence table first.
            std::vector<double> sum(npoints * ncomp, 0.0);
            std::vector<double> wsum(npoints * ncomp, 0.0);
            std::vector<std::int64_t> nodes;
            for (std::size_t b = 0; b < nblocks; ++b) {
                const auto cb = rMesh.Cells(b);
                const detail::DoubleView src(rMesh.CellData(name, b));
                for (std::size_t c = 0; c < cb.NumCells(); ++c) {
                    davg_cell_nodes(cb, c, nodes);
                    const double w = weights[b][c];
                    for (std::size_t k = 0; k < ncomp; ++k) {
                        const double v = src[c * ncomp + k];
                        if (!std::isfinite(v))
                            continue;
                        for (std::int64_t node : nodes) {
                            if (node < 0 || static_cast<std::size_t>(node) >= npoints)
                                continue;
                            const std::size_t idx = static_cast<std::size_t>(node) * ncomp + k;
                            sum[idx] += w * v;
                            wsum[idx] += w;
                        }
                    }
                }
            }
            for (std::size_t i = 0; i < npoints * ncomp; ++i)
                pdst[i] = wsum[i] > 0.0 ? sum[i] / wsum[i] : std::nan("");
            for (std::size_t i = 0; i < npoints * ncomp; ++i)
                pdst[i] =
                    davg_apply_nan_policy(pdst[i], rOpts, "point", name, i / (ncomp ? ncomp : 1));
            out.AddPointData(davg_target_name(name, rOpts), std::move(dst));
            continue;
        }
        if (!incidence) {
            std::vector<std::size_t> block_base(nblocks + 1, 0);
            for (std::size_t b = 0; b < nblocks; ++b)
                block_base[b + 1] = block_base[b] + rMesh.Cells(b).NumCells();
            incidence = std::make_unique<DavgIncidence>(davg_incidence(rMesh, block_base, npoints));
            cell_weight.resize(block_base[nblocks]);
            for (std::size_t b = 0; b < nblocks; ++b)
                std::copy(weights[b].begin(), weights[b].end(),
                          cell_weight.begin() + static_cast<std::ptrdiff_t>(block_base[b]));
            cell_block_base = std::move(block_base);
        }
        // The cell values as doubles, block-major (read_double's conversion).
        std::vector<double> values(cell_weight.size() * ncomp);
        for (std::size_t b = 0; b < nblocks; ++b) {
            const detail::DoubleView src(rMesh.CellData(name, b));
            const std::size_t n = (cell_block_base[b + 1] - cell_block_base[b]) * ncomp;
            double* to = values.data() + cell_block_base[b] * ncomp;
            parallel_for_bw(n, [&](std::size_t i) { to[i] = src[i]; });
        }

        // --- accumulation, gathered per point: each point adds its incident
        // --- cells' weighted values in (block, cell, node) order -- the order
        // --- the serial scatter added them in (FP add is not associative) --
        // --- so the sums are bit-identical on every backend and thread count.
        const DavgIncidence& inc = *incidence;
        parallel_for(npoints, [&](std::size_t pnt) {
            for (std::size_t k = 0; k < ncomp; ++k) {
                double sum = 0.0;
                double wsum = 0.0;
                for (std::uint64_t j = inc.mStart[pnt]; j < inc.mStart[pnt + 1]; ++j) {
                    const std::uint64_t gc = inc.mCellOf[j];
                    const double v = values[gc * ncomp + k];
                    if (!std::isfinite(v))
                        continue;
                    const double w = cell_weight[gc];
                    sum += w * v;
                    wsum += w;
                }
                pdst[pnt * ncomp + k] = wsum > 0.0 ? sum / wsum : std::nan("");
            }
        });
        for (std::size_t i = 0; i < npoints * ncomp; ++i)
            pdst[i] = davg_apply_nan_policy(pdst[i], rOpts, "point", name, i / (ncomp ? ncomp : 1));

        out.AddPointData(davg_target_name(name, rOpts), std::move(dst));
    }
    return out;
}

}  // namespace meshioplusplus

File: src/python/meshioplusplus/_data_average.py
"""Move data between the point and cell locations by averaging.

A dependency-free mesh *operation* (not a file format). The geometry is never
modified; only data arrays are added.

``point_data_to_cell_data`` gives each cell the mean of the values at its own
nodes. ``cell_data_to_point_data`` gives each point the mean of the values on
the cells incident to it, optionally weighted by each cell's |measure| (area for
a 2D cell, volume for a 3D one).

Both work component-wise, so scalar, vector and tensor arrays behave the same.
Because a mean is not an integer, the output is **always** ``float64``.

Non-finite values never contribute to an average; a point or cell with no finite
contribution yields NaN. See :mod:`meshioplusplus._data_common` for the policy.

Public API:
    point_data_to_cell_data
    cell_data_to_point_data
"""

from __future__ import annotations

import copy

import numpy as np

from ._data_common import normalize_location, require_key

#: Outward-wound corner-only face tables, transcribed verbatim (winding
#: included) from `detail/cell_faces.cpp`'s C++ tables -- the authoritative,
#: Newell-normal-gtested source. This table's own consistency IS the
#: correctness of `_cell_measures`' divergence-theorem volume below: the
#: origin-tetrahedra decomposition it uses only gives a translation-invariant
#: answer when every face of the closed surface is wound outward the same
#: way, so a single inward face here would silently make the "volume" of a
#: cell depend on where the cell sits in space rather than on its shape --
#: exactly the defect `TranslatedCellMeasureIsInvariant` below pins.
_FACES = {
    "tetra": [(0, 1, 3), (1, 2, 3), (2, 0, 3), (0, 2, 1)],
    "hexahedron": [
        (0, 4, 7, 3),
        (1, 2, 6, 5),
        (0, 1, 5, 4),
        (3, 7, 6, 2),
        (0, 3, 2, 1),
        (4, 5, 6, 7),
    ],
    "wedge": [(0, 2, 1), (3, 4, 5), (0, 1, 4, 3), (1, 2, 5, 4), (2, 0, 3, 5)],
    "pyramid": [(0, 3, 2, 1), (0, 1, 4), (1, 2, 4), (2, 3, 4), (3, 0, 4)],
}

_CORNERS = {
    "line": 2,
    "triangle": 3,
    "triangle6": 3,
    "quad": 4,
    "quad8": 4,
    "quad9": 4,
    "tetra": 4,
    "tetra10": 4,
    "hexahedron": 8,
    "hexahedron20": 8,
    "hexahedron27": 8,
    "wedge": 6,
    "wedge15": 6,
    "pyramid": 5,
    "pyramid13": 5,
}

_DIM = {
    "line": 1,
    "triangle": 2,
    "triangle6": 2,
    "quad": 2,
    "quad8": 2,
    "quad9": 2,
    "tetra": 3,
    "tetra10": 3,
    "hexahedron": 3,
    "hexahedron20": 3,
    "hexahedron27": 3,
    "wedge": 3,
    "wedge15": 3,
    "pyramid": 3,
    "pyramid13": 3,
}


def _pad3(points):
    """Points padded to three columns, so 2D meshes work unchanged."""
    p = np.asarray(points, dtype=float)
    if p.shape[1] >= 3:
        return p[:, :3]
    return np.hstack([p, np.zeros((p.shape[0], 3 - p.shape[1]))])


def _cell_measures(mesh, block):
    """|measure| of every cell in ``block``, or ones where undefined."""
    ctype = block.type.split("_")[0] if "_" in block.type else block.type
    data = block.data
    if isinstance(data, list) or ctype.startswith(("polygon", "polyhedron")):
        return None
    conn = np.asarray(data)
    corners = _CORNERS.get(ctype)
    dim = _DIM.get(ctype)
    if corners is None or dim is None:
        return None
    pts = _pad3(mesh.points)[conn[:, :corners]]
    if dim == 1:
        return np.linalg.norm(pts[:, 1] - pts[:, 0], axis=1)
    if dim == 2:
        # Newell normal, giving the unsigned polygon area.
        s = np.zeros((pts.shape[0], 3))
        for i in range(corners):
            s += np.cross(pts[:, i], pts[:, (i + 1) % corners])
        return 0.5 * np.linalg.norm(s, axis=1)
    faces = _FACES.get(ctype)
    if faces is None:
        return None
    vol6 = np.zeros(pts.shape[0])
    for face in faces:
        a = pts[:, face[0]]
        for i in range(1, len(face) - 1):
            vol6 += np.einsum(
                "ij,ij->i", a, np.cross(pts[:, face[i]], pts[:, face[i + 1]])
            )
    return np.abs(vol6 / 6.0)


def _cell_node_lists(block):
    """The node ids of each cell, covering ragged and polyhedron blocks."""
    data = block.data
    if block.type.startswith("polyhedron"):
        out = []
        for cell in data:
            seen = []
            marker = set()
            for face in cell:
                for n in face:
                    if n not in marker:
                        marker.add(n)
                        seen.append(int(n))
            out.append(np.asarray(seen, dtype=np.int64))
        return out
    if isinstance(data, list):
        return [np.asarray(row, dtype=np.int64) for row in data]
    conn = np.asarray(data)
    return [conn[i].astype(np.int64) for i in range(conn.shape[0])]


def _apply_nan_policy(values, nan_policy, nan_replacement, name):
    """Apply the shared NaN policy to a produced array."""
    if nan_policy == "ignore":
        return values
    bad = ~np.isfinite(values)
    if not bad.any():
        return values
    if nan_policy == "replace":
        out = values.copy()
        out[bad] = nan_replacement
        return out
    if nan_policy == "fail":
        idx = int(np.argmax(bad.ravel()))
        raise ValueError(
            f"meshio++: data_average: non-finite value at index {idx} of '{name}'"
        )
    raise ValueError(
        f"meshio++: unknown NaN policy '{nan_policy}' "
        "(expected 'ignore', 'replace' or 'fail')"
    )


def _to_cell_py(mesh, names, prefix, suffix, overwrite, nan_policy, nan_replacement):
    """Pure-Python reference for :func:`point_data_to_cell_data`."""
    out = copy.deepcopy(mesh)
    selected = list(names) if names else sorted(mesh.point_data.keys())
    for name in selected:
        require_key(mesh, "point_data", name)
    npoints = len(mesh.points)
    for name in selected:
        target = f"{prefix}{name}{suffix}"
        if not overwrite and target in out.cell_data:
            raise ValueError(
                f"meshio++: data_average: cell_data '{target}' already exists "
                "(pass overwrite=True to replace it)"
            )
        src = np.asarray(mesh.point_data[name], dtype=float)
        ncomp = 1 if src.ndim < 2 else int(np.prod(src.shape[1:]))
        flat = src.reshape(npoints, ncomp)
        blocks = []
        for block in mesh.cells:
            nodes = _cell_node_lists(block)
            res = np.full((len(nodes), ncomp), np.nan)
            for c, ids in enumerate(nodes):
                ids = ids[(ids >= 0) & (ids < npoints)]
                if ids.size == 0:
                    continue
                vals = flat[ids]
                mask = np.isfinite(vals)
                counts = mask.sum(axis=0)
                sums = np.where(mask, vals, 0.0).sum(axis=0)
                with np.errstate(invalid="ignore", divide="ignore"):
                    res[c] = np.where(counts > 0, sums / counts, np.nan)
            res = _apply_nan_policy(res, nan_policy, nan_replacement, name)
            blocks.append(res.reshape(len(nodes)) if ncomp == 1 else res)
        out.cell_data[target] = blocks
    return out


def _to_point_py(
    mesh, names, weight, prefix, suffix, overwrite, nan_policy, nan_replacement
):
    """Pure-Python reference for :func:`cell_data_to_point_data`."""
    out = copy.deepcopy(mesh)
    selected = list(names) if names else sorted(mesh.cell_data.keys())
    for name in selected:
        require_key(mesh, "cell_data", name)
    npoints = len(mesh.points)

    weights = []
    for block in mesh.cells:
        n = len(block.data)
        if weight == "measure":
            w = _cell_measures(mesh, block)
            if w is None:
                w = np.ones(n)
            w = np.where(np.isfinite(w) & (w > 0), w, 1.0)
        else:
            w = np.ones(n)
        weights.append(w)

    for name in selected:
        target = f"{prefix}{name}{suffix}"
        if not overwrite and target in out.point_data:
            raise ValueError(
                f"meshio++: data_average: point_data '{target}' already exists "
                "(pass overwrite=True to replace it)"
            )
        blocks = mesh.cell_data[name]
        if len(blocks) != len(mesh.cells):
            raise ValueError(
                f"meshio++: data_average: cell_data '{name}' has {len(blocks)} "
                f"block(s) but the mesh has {len(mesh.cells)} cell block(s)"
            )
        first = np.asarray(blocks[0], dtype=float)
        ncomp = 1 if first.ndim < 2 else int(np.prod(first.shape[1:]))
        acc = np.zeros((npoints, ncomp))
        wacc = np.zeros((npoints, ncomp))
        for b, block in enumerate(mesh.cells):
            vals = np.asarray(blocks[b], dtype=float).reshape(-1, ncomp)
            nodes = _cell_node_lists(block)
            for c, ids in enumerate(nodes):
                ids = ids[(ids >= 0) & (ids < npoints)]
                if ids.size == 0:
                    continue
                v = vals[c]
                finite = np.isfinite(v)
                if not finite.any():
                    continue
                w = weights[b][c]
                contrib = np.where(finite, w * v, 0.0)
                wcontrib = np.where(finite, w, 0.0)
                np.add.at(acc, ids, contrib)
                np.add.at(wacc, ids, wcontrib)
        with np.errstate(invalid="ignore", divide="ignore"):
            res = np.where(wacc > 0, acc / np.where(wacc > 0, wacc, 1.0), np.nan)
        res = _apply_nan_policy(res, nan_policy, nan_replacement, name)
        out.point_data[target] = res.reshape(npoints) if ncomp == 1 else res
    return out


def point_data_to_cell_data(
    mesh,
    keys=None,
    prefix: str = "",
    suffix: str = "",
    overwrite: bool = True,
    nan_policy: str = "ignore",
    nan_replacement: float = 0.0,
):
    """Average ``point_data`` onto the cells.

    Each cell's value is the mean over its own nodes.

    Args:
        mesh: the source mesh (unmodified).
        keys: names to convert; ``None`` means every ``point_data`` array.
        prefix: output name is ``prefix + name + suffix``.
        suffix: see ``prefix``.
        overwrite: whether an existing target name may be replaced.
        nan_policy: ``"ignore"``, ``"replace"`` or ``"fail"``.
        nan_replacement: value used when ``nan_policy`` is ``"replace"``.

    Returns:
        A new mesh carrying the produced ``cell_data``.
    """
    names = list(keys) if keys else []
    try:
        from . import _core
    except Exception:
        return _to_cell_py(
            mesh, names, prefix, suffix, overwrite, nan_policy, nan_replacement
        )
    try:
        out = _core.point_data_to_cell_data(
            mesh, names, prefix, suffix, overwrite, nan_policy, nan_replacement
        )
    except ValueError:
        raise
    except Exception:
        return _to_cell_py(
            mesh, names, prefix, suffix, overwrite, nan_policy, nan_replacement
        )
    _carry_sets(mesh, out)
    return out


def cell_data_to_point_data(
    mesh,
    keys=None,
    weighted: bool = False,
    prefix: str = "",
    suffix: str = "",
    overwrite: bool = True,
    nan_policy: str = "ignore",
    nan_replacement: float = 0.0,
):
    """Average ``cell_data`` onto the points.

    Each point's value is the mean over the cells incident to it, optionally
    weighted by each cell's |measure|.

    Args:
        mesh: the source mesh (unmodified).
        keys: names to convert; ``None`` means every ``cell_data`` array.
        weighted: weight by cell measure instead of counting cells equally.
        prefix: output name is ``prefix + name + suffix``.
        suffix: see ``prefix``.
        overwrite: whether an existing target name may be replaced.
        nan_policy: ``"ignore"``, ``"replace"`` or ``"fail"``.
        nan_replacement: value used when ``nan_policy`` is ``"replace"``.

    Returns:
        A new mesh carrying the produced ``point_data``.
    """
    names = list(keys) if keys else []
    weight = "measure" if weighted else "uniform"
    try:
        from . import _core
    except Exception:
        return _to_point_py(
            mesh, names, weight, prefix, suffix, overwrite, nan_policy, nan_replacement
        )
    try:
        out = _core.cell_data_to_point_data(
            mesh, names, weight, prefix, suffix, overwrite, nan_policy, nan_replacement
        )
    except ValueError:
        raise
    except Exception:
        return _to_point_py(
            mesh, names, weight, prefix, suffix, overwrite, nan_policy, nan_replacement
        )
    _carry_sets(mesh, out)
    return out


def _carry_sets(source, target) -> None:
    """Copy sets across; they index geometry, which is never modified here."""
    for attr in ("point_sets", "cell_sets"):
        value = getattr(source, attr, None)
        if value:
            setattr(target, attr, copy.deepcopy(value))


__all__ = ["point_data_to_cell_data", "cell_data_to_point_data", "normalize_location"]
